### DISCUSSION TITLE
Renovate not respecting Python '<=' constraint correctly for Poetry

### DISCUSSION THREAD

**Discussion by jvmdc** [renovatebot/renovate#31474] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

When using a less-than-or-equal ('<=') version constraint for the Python version in a `pyproject.toml` file with Poetry, Renovate can pick up the wrong Python version and fail to update the lock file when updating other packages.

Repository with issue: https://github.com/jvmdc/PythonProject1 - See PR https://github.com/jvmdc/PythonProject1/pull/3. Renovate has used Python 3.12.6, but maximum should be 3.12(.0?)

While the intent with the version constraint in the above repository was to include 3.12.x versions, this is not how the versioning constraint works in Poetry apparently - See also https://github.com/python-poetry/poetry/issues/8418#issuecomment-1853994659

To limit to 3.12.x, `<3.13` can be used successfully instead as done in https://github.com/jvmdc/PythonProject2 - PR equivalent to the above one is here with no issues in lock-file updating: https://github.com/jvmdc/PythonProject2/pull/2

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: Updating poetry.lock (branch="renovate/pytest-7.x")
DEBUG: Using python version from pyproject.toml (branch="renovate/pytest-7.x")
DEBUG: Using poetry version 1.7.1 from poetry.lock header (branch="renovate/pytest-7.x")
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/renovate/cache/containerbase (branch="renovate/pytest-7.x")
DEBUG: Using containerbase dynamic installs (branch="renovate/pytest-7.x")
DEBUG: Resolved stable matching version (branch="renovate/pytest-7.x")
{
  "toolName": "python"
  "constraint": ">=3.9,<=3.12"
  "resolvedVersion": "3.12.6"
}

DEBUG: Executing command (branch="renovate/pytest-7.x")
{
  "command": "install-tool python 3.12.6"
}

DEBUG: exec completed (branch="renovate/pytest-7.x")
{
  "durationMs": 606
  "stdout": "tool python v3.12.6 is already installed\ntool is already linked: python v3.12.6\n[07:55:30.345] INFO (337): Installing tool python@3.12.6...\n[07:55:30.428] INFO (337): Install tool python succeeded in 111ms.\n"
  "stderr": ""
}

DEBUG: Executing command (branch="renovate/pytest-7.x")
{
  "command": "install-tool poetry 1.7.1"
}

DEBUG: exec completed (branch="renovate/pytest-7.x")
{
  "durationMs": 1598
  "stdout": "[07:55:30.926] INFO (371): Installing pip poetry@1.7.1...\n[07:55:30.947] INFO (371): tool already installed\n    tool: \"poetry\"\nPoetry (version 1.7.1)\n[07:55:31.540] INFO (371): Install tool poetry succeeded in 654ms.\n"
  "stderr": ""
}

DEBUG: Executing command (branch="renovate/pytest-7.x")
{
  "command": "poetry update --lock --no-interaction pytest"
}

DEBUG: rawExec err (branch="renovate/pytest-7.x")
{
  "err": {
    "cmd": "/bin/sh -c poetry update --lock --no-interaction pytest",
    "stderr": "\nCurrent Python version (3.12.6) is not allowed by the project (>=3.9,<=3.12).\nPlease change python executable via the \"env use\" command.\n",
    "stdout": "",
    "options": {
      "cwd": "/tmp/renovate/repos/github/jvmdc/PythonProject1",
      "encoding": "utf-8",
      "env": {
        "GIT_CONFIG_KEY_0": "url.https://**redacted**@github.com/.insteadOf",
        "GIT_CONFIG_VALUE_0": "ssh://**redacted**@github.com/",
        "GIT_CONFIG_KEY_1": "url.https://**redacted**@github.com/.insteadOf",
        "GIT_CONFIG_VALUE_1": "git@github.com:",
        "GIT_CONFIG_KEY_2": "url.https://**redacted**@github.com/.insteadOf",
        "GIT_CONFIG_VALUE_2": "https://github.com/",
        "GIT_CONFIG_COUNT": "3",
        "PIP_CACHE_DIR": "/tmp/renovate/cache/others/pip",
        "HOME": "/home/ubuntu",
        "PATH": "/home/ubuntu/.local/bin:/home/ubuntu/bin:/home/ubuntu/.local/bin:/home/ubuntu/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "CONTAINERBASE_CACHE_DIR": "/tmp/renovate/cache/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "name": "ExecError",
    "message": "Command failed: poetry update --lock --no-interaction pytest\n\nCurrent Python version (3.12.6) is not allowed by the project (>=3.9,<=3.12).\nPlease change python executable via the \"env use\" command.\n",
    "stack": "ExecError: Command failed: poetry update --lock --no-interaction pytest\n\nCurrent Python version (3.12.6) is not allowed by the project (>=3.9,<=3.12).\nPlease change python executable via the \"env use\" command.\n\n    at ChildProcess.<anonymous> (/usr/local/renovate/lib/util/exec/common.ts:101:11)\n    at ChildProcess.emit (node:events:531:35)\n    at ChildProcess.emit (node:domain:488:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:294:12)"
  }
  "durationMs": 1146
}

DEBUG: Failed to update poetry.lock file (branch="renovate/pytest-7.x")
{
  "err": {
    "cmd": "/bin/sh -c poetry update --lock --no-interaction pytest",
    "stderr": "\nCurrent Python version (3.12.6) is not allowed by the project (>=3.9,<=3.12).\nPlease change python executable via the \"env use\" command.\n",
    "stdout": "",
    "options": {
      "cwd": "/tmp/renovate/repos/github/jvmdc/PythonProject1",
      "encoding": "utf-8",
      "env": {
        "GIT_CONFIG_KEY_0": "url.https://**redacted**@github.com/.insteadOf",
        "GIT_CONFIG_VALUE_0": "ssh://**redacted**@github.com/",
        "GIT_CONFIG_KEY_1": "url.https://**redacted**@github.com/.insteadOf",
        "GIT_CONFIG_VALUE_1": "git@github.com:",
        "GIT_CONFIG_KEY_2": "url.https://**redacted**@github.com/.insteadOf",
        "GIT_CONFIG_VALUE_2": "https://github.com/",
        "GIT_CONFIG_COUNT": "3",
        "PIP_CACHE_DIR": "/tmp/renovate/cache/others/pip",
        "HOME": "/home/ubuntu",
        "PATH": "/home/ubuntu/.local/bin:/home/ubuntu/bin:/home/ubuntu/.local/bin:/home/ubuntu/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "CONTAINERBASE_CACHE_DIR": "/tmp/renovate/cache/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "name": "ExecError",
    "message": "Command failed: poetry update --lock --no-interaction pytest\n\nCurrent Python version (3.12.6) is not allowed by the project (>=3.9,<=3.12).\nPlease change python executable via the \"env use\" command.\n",
    "stack": "ExecError: Command failed: poetry update --lock --no-interaction pytest\n\nCurrent Python version (3.12.6) is not allowed by the project (>=3.9,<=3.12).\nPlease change python executable via the \"env use\" command.\n\n    at ChildProcess.<anonymous> (/usr/local/renovate/lib/util/exec/common.ts:101:11)\n    at ChildProcess.emit (node:events:531:35)\n    at ChildProcess.emit (node:domain:488:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:294:12)"
  }
}


```

</details>

**Comment by jvmdc** [comment]
FYI for the reproduction repositories made public here I used the Renovate-hosted app, but we're using self-hosted internally - Same issue seen in both cases.

**Comment by rarkins** [comment]
This can also be reproduced using unit tests. I added three, and the last one failed like you described (`matches('3.21.1', '<=3.12')` returns true instead of false). The third one here should be a tick if it reported `false` correctly:

    ✓ matches("3.11.9", "<= 3.12") === "true"
    ✓ matches("3.12.0", "<= 3.12") === "true"
    ✕ matches("3.12.1", "<= 3.12") === "false"

**Reply by jvmdc** [reply]
@rarkins poetry-core actually has a set of test cases themselves (https://github.com/python-poetry/poetry-core/blob/8c878fd4a81b526dc8d91fcd5f7406ece37d9202/tests/constraints/version/test_version_range.py#L451-L629) that fit nicely into the `matches` test. 

I extracted them and added them as test cases (for the `matches` test as well as testing both range and version in the `isValid` test). These are the ones that failed - I've filtered the ones using Poetry's negation operator out as it's explicitly mentioned to be unsupported by Renovate's current implementation in the `poetry2npm` function:

```
✕ isValid("==2!1.*") === true (1 ms)
✕ isValid("==2!1.0") === true
✕ isValid("==2.0.0") === true
✕ isValid("==2.0+deadbeef") === true (1 ms)
✕ isValid("~=2!1.0") === true
✕ isValid("1.1.post1") === true
✕ isValid("2!1.0") === true
✕ isValid("2.0.dev1") === true
✕ isValid("2.0.post1.dev1") === true
✕ isValid("2.0.post1") === true
✕ isValid("2.0+deadbeef.0") === true
✕ isValid("2.0+deadbeef") === true
✕ isValid("2.0+local.version") === true
✕ isValid("2.0a1.dev1") === true
✕ isValid("2.0a1.post1") === true (1 ms)
✕ isValid("2.0a1") === true
✕ isValid("2.0b1.dev1") === true
✕ isValid("2.0b1.post1") === true
✕ isValid("2.0b1") === true
✕ isValid("2.0b2.dev1") === true
✕ isValid("2.0c1.post1.dev1") === true
✕ isValid("2.0c1") === true
✕ isValid("2.0rc1") === true (1 ms)
✕ isValid("2.dev1") === true
✕ isValid("2.post1.dev1") === true
✕ isValid("2.post1") === true
✕ isValid("2a1.post1") === true
✕ isValid("2a1") === true
✕ isValid("2b1.dev1") === true
✕ isValid("2b1") === true
✕ isValid("2c1.post1.dev1") === true (1 ms)
✕ isValid("2c1") === true
✕ isValid("2rc1") === true
✕ matches("3.12.1", "<= 3.12") === "false"
✕ matches("2.0", "==2.0.0") === "true"
✕ matches("2.0+deadbeef", "==2.0.0") === "true"
✕ matches("2.0+deadbeef", "==2+deadbeef") === "true"
✕ matches("2.0+deadbeef", "==2.0+deadbeef") === "true" (1 ms)
✕ matches("2.0+deadbeef", "==2.0.0+deadbeef") === "true"
✕ matches("2.0+deadbeef.0", "==2.0.0+deadbeef.00") === "true"
✕ matches("2.dev1", "==2.*") === "true" (1 ms)
✕ matches("2a1", "==2.*") === "true"
✕ matches("2a1.post1", "==2.*") === "true" (1 ms)
✕ matches("2b1", "==2.*") === "true"
✕ matches("2b1.dev1", "==2.*") === "true"
✕ matches("2c1", "==2.*") === "true"
✕ matches("2c1.post1.dev1", "==2.*") === "true" (1 ms)
✕ matches("2rc1", "==2.*") === "true"
✕ matches("2.0.post1", "==2.0.post1.*") === "true" (1 ms)
✕ matches("2.0.post1.dev1", "==2.0.post1.*") === "true"
✕ matches("2.0.post1", ">=2") === "true"
✕ matches("2.0.post1.dev1", ">=2") === "true" (1 ms)
✕ matches("2.0.dev1", "<=2") === "true" (1 ms)
✕ matches("2.0a1", "<=2") === "true"
✕ matches("2.0a1.dev1", "<=2") === "true"
✕ matches("2.0b1", "<=2") === "true"
✕ matches("2.0b1.post1", "<=2") === "true"
✕ matches("2.0c1", "<=2") === "true" (1 ms)
✕ matches("2.0c1.post1.dev1", "<=2") === "true"
✕ matches("2.0rc1", "<=2") === "true"
✕ matches("2.0.1", ">2") === "true"
✕ matches("2.1.post1", ">2") === "true" (1 ms)
✕ matches("2.1+local.version", ">2") === "true"
✕ matches("2.0.dev0", "<2.1") === "true" (1 ms)
✕ matches("1.1", "~=1.0") === "true"
✕ matches("1.9999999", "~=1.0") === "true" (1 ms)
✕ matches("1.1", "~=1.0a1") === "true"
✕ matches("2!1.0", "~=2!1.0") === "true"
✕ matches("2!1.0", "==2!1.*") === "true"
✕ matches("2!1.0", "==2!1.0") === "true"
✕ matches("1.0", "<=2!0.1") === "true"
✕ matches("2!1.0", ">=2.0") === "true"
✕ matches("1.0", "<2!0.1") === "true"
✕ matches("2!1.0", ">2.0") === "true"
✕ matches("2.0.5", ">2.0dev") === "true" (1 ms)
✕ matches("2.1", "==2") === "false"
✕ matches("2!1.0", "~=1.0") === "false"
✕ matches("2!1.0", "==1.0") === "false" (1 ms)
✕ matches("2!1.0", "==1.*") === "false"
```

The nice thing about them having an almost identical test is that we can also go the other way - I added the tests in Renovate's `matches` test for Poetry to poetry-core's test suite and a few are failing: 
```
FAILED tests/constraints/version/test_version_range.py::test_specifiers[1.9.4-===1.9.4-True] - poetry.core.constraints.version.exceptions.ParseConstraintError: Could not parse version constraint: ===1.9.4
FAILED tests/constraints/version/test_version_range.py::test_specifiers[1.9.4-==1.9-True] - AssertionError: assert False
FAILED tests/constraints/version/test_version_range.py::test_specifiers[1.9.4-===1.9.3-False] - poetry.core.constraints.version.exceptions.ParseConstraintError: Could not parse version constraint: ===1.9.3
```
In other words, `===` does not exist as a range indicator/constraint in Poetry and `✓ matches("1.9.4", "==1.9") === "true" ` is wrong (which makes sense - per PEP440, only substitution happening by default for `==` is zero padding to ensure versions are the same length for comparison, i.e. `1.9` becomes `1.9.0` when comparing to `1.9.4`)

**Reply by rarkins** [reply]
PRs welcome

**Reply by HonkingGoose** [reply]
Related PR, which the author closed due to the risk of breaking things:

- https://github.com/renovatebot/renovate/pull/31414

It looks like we have some good tests in this discussion. With better tests, maybe the change is safe to do now?

**Reply by jvmdc** [reply]
> PRs welcome

Yeah, no worries. I have been working on it a bit, but haven't gotten far - This is just as much findings for me to remember as for anyone else who'd like to give it a shot :)
